crates/kernel/src/builtins/bf_list_sets.rs

```diff
@@ -598,62 +598,83 @@ fn bf_slice(bf_args: &mut BfCallState<'_>) -> Result<BfRet, BfErr> {
 
     match collection.variant() {
         Variant::List(list) => {
+            // Ensure we have a list of lists or maps
+            if list.is_empty() {
+                return Ok(Ret(v_empty_list()));
+            }
+            
+            let first_item = list.index(0).map_err(BfErr::Code)?;
+            if !matches!(first_item.variant(), Variant::List(_) | Variant::Map(_)) {
+                return Err(BfErr::Code(E_TYPE));
+            }
+            
             match index.variant() {
-                // Case 1: List + Integer index
+                // Case 1: List of lists + Integer index
+                // This handles: slice({{1,2,3},{4,5,6}}, 2) => {2, 5}
+                // For each sublist in the input list, extract the element at position 'idx'
+                // and return a list of these elements
                 Variant::Int(idx) => {
                     let idx = *idx as usize;
+                    let mut result = Vec::with_capacity(list.len());
 
-                    // Check if this is a list of lists
-                    let first_item = list.index(0).map_err(BfErr::Code)?;
-                    if let Variant::List(_) = first_item.variant() {
-                        // This is a list of lists, extract the idx-th element from each sublist
-                        let mut result = Vec::with_capacity(list.len());
-
-                        for item in list.iter() {
-                            if let Variant::List(sublist) = item.variant() {
-                                if idx < 1 || idx > sublist.len() {
-                                    return Err(BfErr::Code(E_RANGE));
-                                }
-                                // MOO is 1-indexed, so subtract 1
-                                result.push(sublist.index(idx - 1).map_err(BfErr::Code)?);
-                            } else {
-                                return Err(BfErr::Code(E_TYPE));
+                    for item in list.iter() {
+                        if let Variant::List(sublist) = item.variant() {
```

Comment:
Nit: one way to write this to avoid this deep of nesting is to flip it around:

```rust
let Variant::List(sublist) = item.variant() else {
  return Err(BfErr::Code(E_TYPE));
};

if idx < 1 ...
```

It helps a lot with keeping matches and if blocks from getting too deeply nested